core/src/main/java/ai/timefold/solver/core/impl/domain/entity/descriptor/EntityDescriptor.java

```diff
@@ -330,11 +331,21 @@ private void processSorting(PlanningEntity entityAnnotation) {
     private void processValueRangeProviderAnnotation(DescriptorPolicy descriptorPolicy, Member member) {
         if (((AnnotatedElement) member).isAnnotationPresent(ValueRangeProvider.class)) {
             var memberAccessor = descriptorPolicy.getMemberAccessorFactory().buildAndCacheMemberAccessor(member,
-                    FIELD_OR_READ_METHOD, ValueRangeProvider.class, descriptorPolicy.getDomainAccessType());
+                    FIELD_OR_READ_METHOD_WITH_PARAMETER, ValueRangeProvider.class, descriptorPolicy.getDomainAccessType());
+            assertGetterParameterType(memberAccessor);
             descriptorPolicy.addFromEntityValueRangeProvider(memberAccessor);
         }
     }
 
+    private void assertGetterParameterType(MemberAccessor memberAccessor) {
+        if (memberAccessor.acceptsParameter()
+                && !memberAccessor.getGetterMethodParameterType().equals(getSolutionDescriptor().getSolutionClass())) {
+            throw new IllegalStateException("The parameter type (%s) of the method (%s) must match the solution (%s)."
+                    .formatted(memberAccessor.getGetterMethodParameterType().getTypeName(), memberAccessor.getName(),
+                            getSolutionDescriptor().getSolutionClass().getSimpleName()));
```

Comment:
The solution class is a user type, therefore we do not give a simple name, but a fully qualified name. Applies to both the method parameter type and to the solution class itself.